File: src/main/java/com/github/sorusclient/client/adapter/IRenderer.java
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.adapter;

import com.github.sorusclient.client.util.Color;
import org.apache.commons.io.IOUtils;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

public interface IRenderer {

    void draw(RenderBuffer buffer);
    void setColor(Color color);

    default void drawRectangle(double x, double y, double width, double height, Color color) {
        this.drawRectangle(x, y, width, height, 0.0, color);
    }
    default void drawRectangle(double x, double y, double width, double height, double cornerRadius, Color color) {
        this.drawRectangle(x, y, width, height, cornerRadius, color, color, color, color);
    }
    void drawRectangle(double x, double y, double width, double height, double cornerRadius, Color topLeftColor, Color bottomLeftColor, Color bottomRightColor, Color topRightColor);

    default void drawRectangleBorder(double x, double y, double width, double height, double thickness, Color color) {
        drawRectangleBorder(x, y, width, height, 0.0, thickness, color);
    }

    void drawRectangleBorder(double x, double y, double width, double height, double cornerRadius, double thickness, Color color);

    default void drawImage(String id, double x, double y, double width, double height, Color color) {
        drawImage(id, x, y, width, height, true, color);
    }

    default void drawImage(String id, double x, double y, double width, double height, boolean antialias, Color color) {
        drawImage(id, x, y, width, height, 0.0, antialias, color);
    }

    default void drawImage(String id, double x, double y, double width, double height, double cornerRadius, boolean antialias, Color color) {
        drawImage(id, x, y, width, height, cornerRadius, 0.0, 0.0, 1.0, 1.0, antialias, color);
    }

    void drawImage(String id, double x, double y, double width, double height, double cornerRadius, double textureX, double textureY, double textureWidth, double textureHeight, boolean antialias, Color color);

    default void createTexture(String path) {
        var resource = IRenderer.class.getClassLoader().getResource(path);
        if (resource != null) {
            createTexture(path, resource);
        }
    }

    default void createTexture(String id, URL url) {
        try {
            createTexture(id, url.openStream(), true);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    default void createTexture(String id, InputStream inputStream, boolean antialias) {
        try {
            createTexture(id, IOUtils.toByteArray(inputStream), antialias);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    void createTexture(String id, byte[] bytes, boolean antialias);

    void scissor(double x, double y, double width, double height);
    void endScissor();

    IFontRenderer getFontRenderer(String id);

    void drawText(String id, String text, double x, double y, double scale, Color color);
    double getTextWidth(String fontId, String text);
    double getTextHeight(String fontId);

    default void createFont(String path) {
        createFont(path, IRenderer.class.getClassLoader().getResource(path));
    }
    default void createFont(String id, URL url) {
        try {
            createFont(id, url.openStream());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    void createFont(String id, InputStream inputStream);

    void loadBlur();
    void unloadBlur();

}

File: src/main/kotlin/com/github/sorusclient/client/adapter/IRenderer.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.adapter

import com.github.sorusclient.client.util.Color
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.io.InputStream
import java.net.URL
import java.util.*

interface IRenderer {
    fun draw(buffer: RenderBuffer)
    fun setColor(color: Color)
    fun setLineThickness(thickness: Double)
    fun drawRectangle(x: Double, y: Double, width: Double, height: Double, color: Color) {
        this.drawRectangle(x, y, width, height, 0.0, color)
    }
    fun drawRectangle(x: Double, y: Double, width: Double, height: Double, cornerRadius: Double, color: Color) {
        this.drawRectangle(x, y, width, height, cornerRadius, color, color, color, color)
    }

    fun drawRectangle(
        x: Double,
        y: Double,
        width: Double,
        height: Double,
        cornerRadius: Double,
        topLeftColor: Color,
        bottomLeftColor: Color,
        bottomRightColor: Color,
        topRightColor: Color
    )

    fun drawRectangleBorder(x: Double, y: Double, width: Double, height: Double, thickness: Double, color: Color) {
        drawRectangleBorder(x, y, width, height, 0.0, thickness, color)
    }

    fun drawRectangleBorder(x: Double, y: Double, width: Double, height: Double, cornerRadius: Double, thickness: Double, color: Color)

    fun drawImage(id: String, x: Double, y: Double, width: Double, height: Double, color: Color) {
        drawImage(id, x, y, width, height, true, color)
    }

    fun drawImage(id: String, x: Double, y: Double, width: Double, height: Double, antialias: Boolean, color: Color) {
        drawImage(id, x, y, width, height, 0.0, antialias, color)
    }

    fun drawImage(id: String, x: Double, y: Double, width: Double, height: Double, cornerRadius: Double, antialias: Boolean, color: Color) {
        drawImage(id, x, y, width, height, cornerRadius, 0.0, 0.0, 1.0, 1.0, antialias, color)
    }

    fun drawImage(id: String, x: Double, y: Double, width: Double, height: Double, cornerRadius: Double, textureX: Double, textureY: Double, textureWidth: Double, textureHeight: Double, antialias: Boolean, color: Color)

    fun createTexture(path: String) {
        IRenderer::class.java.classLoader.getResource(path)?.let { createTexture(path, it) }
    }

    fun createTexture(id: String, url: URL) {
        try {
            createTexture(id, url.openStream(), true)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun createTexture(id: String, inputStream: InputStream, antialias: Boolean) {
        createTexture(id, IOUtils.toByteArray(inputStream), antialias)
    }

    fun createTexture(id: String, bytes: ByteArray, antialias: Boolean)

    fun scissor(x: Double, y: Double, width: Double, height: Double)
    fun endScissor()
    fun getFontRenderer(id: String): IFontRenderer?

    fun drawText(id: String, text: String, x: Double, y: Double, scale: Double, color: Color)
    fun getTextWidth(fontId: String, text: String): Double
    fun getTextHeight(fontId: String): Double

    fun createFont(path: String) {
        createFont(path, Objects.requireNonNull(IRenderer::class.java.classLoader.getResource(path)))
    }

    fun createFont(id: String, url: URL) {
        try {
            createFont(id, url.openStream())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun createFont(id: String, inputStream: InputStream)

    fun loadBlur()
    fun unloadBlur()

}